pdns/dnsdistdist/dnsdist-tcp-downstream.cc

```diff
@@ -304,8 +304,37 @@ IOState TCPConnectionToBackend::sendQuery(std::shared_ptr<TCPConnectionToBackend
   return state;
 }
 
+class HandlingIOGuard
+{
+public:
+  HandlingIOGuard(bool& handlingIO) :
+    d_handlingIO(handlingIO)
+  {
+  }
+  HandlingIOGuard(const HandlingIOGuard&) = delete;
+  HandlingIOGuard(HandlingIOGuard&&) = delete;
+  HandlingIOGuard& operator=(const HandlingIOGuard& rhs) = delete;
+  HandlingIOGuard& operator=(HandlingIOGuard&&) = delete;
+  ~HandlingIOGuard()
+  {
+    d_handlingIO = false;
+  }
+
+private:
+  bool& d_handlingIO;
+};
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
 void TCPConnectionToBackend::handleIO(std::shared_ptr<TCPConnectionToBackend>& conn, const struct timeval& now)
 {
+  if (conn->d_handlingIO) {
+    conn->d_handleSkipped = true;
+    return;
+  }
+
+  conn->d_handlingIO = true;
+  HandlingIOGuard reentryGuard(conn->d_handlingIO);
```

Comment:
You should be able to use:
```suggestion
  dnsdist::tcp::HandlingIOGuard reentryGuard(conn->d_handlingIO);
```
so you don't have to duplicate `HandlingIOGuard`.